sw/device/silicon_creator/lib/drivers/kmac.c

```diff
@@ -0,0 +1,297 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/kmac.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "kmac_regs.h"  // Generated.
+
+enum {
+  /**
+   * Base address of the KMAC hardware MMIO interface.
+   */
+  kBase = TOP_EARLGREY_KMAC_BASE_ADDR,
+  /**
+   * Keccak capacity for SHAKE256.
+   *
+   * See FIPS 202, section 6.2.
+   */
+  kShake256KeccakCapacity = 2 * 256,
+  /**
+   * Keccak rate for SHAKE256 (bits).
+   *
+   * Rate is 1600 - capacity (FIPS 202, section 6.2).
+   */
+  kShake256KeccakRateBits = 1600 - kShake256KeccakCapacity,
+  /**
+   * Keccak rate for SHAKE256 (bytes).
+   */
+  kShake256KeccakRateBytes = kShake256KeccakRateBits / 8,
+  /**
+   * Keccak rate for SHAKE256 (words).
+   */
+  kShake256KeccakRateWords = kShake256KeccakRateBytes / sizeof(uint32_t),
+  /**
+   * Size of one share of the Keccak state.
+   */
+  kStateShareSize = KMAC_STATE_SIZE_BYTES / 2,
+  /**
+   * Address of first share of Keccak state.
+   */
+  kAddrStateShare0 = kBase + KMAC_STATE_REG_OFFSET,
+  /**
+   * Address of second share of Keccak state.
+   */
+  kAddrStateShare1 = kBase + KMAC_STATE_REG_OFFSET + kStateShareSize,
+};
+
+// Double-check that calculated rate is smaller than one share of the state.
+static_assert(kShake256KeccakRateWords <= kStateShareSize,
+              "assert SHAKE256 rate is <= share size");
+
+/**
+ * Polls the KMAC block state until the desired status bit is set.
+ *
+ * If the KMAC block registers an error, this routine exits early and returns
+ * `kErrorKmacInvalidStatus`.
+ *
+ * @param bit_index Bit within the status register to poll.
+ * @return Result of the operation.
+ */
+static rom_error_t poll_state(bitfield_bit32_index_t bit_index) {
+  // The success condition of this function is:
+  //   - The specified bit in the status register is 1, and
+  //   - The error bit in KMAC's INTR_STATE register is 0.
+  //
+  // In order to make fault injection more difficult, we compute these values
+  // in a slightly convoluted way so that skipping any few instructions will
+  // not reach the success condition.
+  uint32_t status = 0;
+  rom_error_t res = launder32(kErrorOk ^ UINT32_MAX);
+  uint32_t is_error = (uint32_t)kHardenedBoolFalse;
+  do {
+    // Read the error bit.
+    uint32_t intr_state = abs_mmio_read32(kBase + KMAC_INTR_STATE_REG_OFFSET);
+    uint32_t err_bit = launder32((uint32_t)bitfield_bit32_read(
```

Comment:
nit: Is `launder32()` required here? `intr_state` is initialized using a `volatile` access and the result of the bit mask is laundered. I guess the compiler can only infer that this is a single bit value, right? 

Also applies to `flag` below.

Reply:
Yes, the reason for the laundering is to ensure the compiler can't infer that the value is single-bit (and therefore won't turn the bitwise not into a branch or something similar). I'll check the disassemblies for this with and without the `launder32` and change it in a follow-up if it's not needed.